### PR TITLE
fix potential inconsistency

### PR DESCRIPTION
**Pull request by zhangsoledad**
base on https://github.com/nervosnetwork/ckb/pull/1587

1. remove `ChainProvider`, it's superfluous.
2. fix access storage directly in many places, it has no consistency guarantee, it's potential problems
3. ~~fix snapshot inconsistency with get_pinned~~

### CODE REVIEW

File: sync/src/tests/sync_shared_state.rs [hunk 1]
```diff
@@ -155,14 +162,15 @@ fn test_switch_invalid_fork() {
     };
 
     // Insert the invalid fork. The fork blocks would not been verified until the fork switches as
-    // the main chain. So `insert_new_block` is ok even for invalid block. And `block_status_map`
+    // the main chain. So .snapshot()`insert_new_block` is ok even for invalid block. And `block_status_map`
```
Comment (u2) [inline sync/src/tests/sync_shared_state.rs]: typo

Comment (zhangsoledad) [reply sync/src/tests/sync_shared_state.rs]: :+1:

File: chain/src/chain.rs [hunk 2]
```diff
@@ -366,9 +365,9 @@ impl ChainService {
             let mut txs_verify_cache = self.shared.lock_txs_verify_cache();
 
             let new_snapshot = self.shared.new_snapshot(
-                tip_header,
+                tip_header.clone(),
                 total_difficulty,
-                epoch,
+                epoch.clone(),
```
[the remark this answers is not in the document]
Comment (zhangsoledad) [reply chain/src/chain.rs]: :+1:

**Comment by quake** [comment]
bors r=quake,u2

**Comment by u2** [comment]
bors r=quake,u2

**Pull request merged**